xchange-kraken/src/main/java/info/bitrich/xchangestream/kraken/KrakenStreamingTradeService.java

```diff
@@ -0,0 +1,221 @@
+    }
+    private String getChannelName(KrakenSubscriptionName subscriptionName) {
+        return subscriptionName.toString();
+    }
+
+    private static class KrakenDtoOrderHolder extends HashMap<String, KrakenOpenOrder> {}
+    private static class KrakenDtoUserTradeHolder extends HashMap<String, KrakenOwnTrade> {}
+
+    private KrakenWebsocketToken renewToken() throws IOException {
+        if (System.currentTimeMillis() >= tokenExpires) {
+            token = rawKrakenAcctService.getKrakenWebsocketToken();
+            tokenExpires = System.currentTimeMillis() + (token.getExpiresInSeconds() * 900L); // 900L instead of a 1000L for 90%
+        }
+        return token;
+    }
+
+    @Override
+    public Observable<Order> getOrderChanges(CurrencyPair currencyPair, Object... args) {
+
+        return Observable.create( source -> {
+            try {
+                String channelName = getChannelName(KrakenSubscriptionName.openOrders);
+
+                streamingService.subscribeChannel(channelName, renewToken().getToken())
+                        .filter(JsonNode::isArray)
+                        .filter(Objects::nonNull)
+                        .map(jsonNode -> jsonNode.get(0))
+                        .map(jsonNode ->
+                                StreamingObjectMapperHelper.getObjectMapper().treeToValue(jsonNode, KrakenDtoOrderHolder[].class))
+                        .doOnError(source::onError)
+                        .forEach(list ->
+                                adaptKrakenOrders(list)
+                                        .forEach(source::onNext));
+
+            } catch (Throwable e) {
+                source.onError(e);
+            }
+        });
+    }
+
+    private List<Order> adaptKrakenOrders(KrakenDtoOrderHolder[] dtoList) {
+        List<Order> result = new ArrayList<>();
+
+        for(KrakenDtoOrderHolder dtoHolder : dtoList) {
+            for (Map.Entry<String, KrakenOpenOrder> dtoOrderEntry : dtoHolder.entrySet()) {
+                String orderId = dtoOrderEntry.getKey();
+                KrakenOpenOrder dto = dtoOrderEntry.getValue();
+
+                Order.OrderStatus status = dto.status == null ? null : KrakenAdapters.adaptOrderStatus(KrakenOrderStatus.fromString(dto.status));
+                KrakenOpenOrder.KrakenDtoDescr descr = dto.descr;
+
+                CurrencyPair pair = descr == null ? null : new CurrencyPair(descr.pair);
+                Order.OrderType side = descr == null ? null : KrakenAdapters.adaptOrderType(KrakenType.fromString(descr.type));
+
+                Date time = dto.opentm == null ? null : new Date((long) (dto.opentm * 1000L));
+                String ordertype = (descr == null || descr.ordertype == null) ? null : descr.ordertype;
+
+                Set<Order.IOrderFlags> flags = adaptFlags(dto.oflags);
+
+                if ("limit".equals(ordertype))
+                    result.add(
+                            new LimitOrder.Builder(side, pair)
```

Comment:
Most of these blocks are duplicated.  `LimitOrder.Builder` extends `Order.Builder` so you can create the builder in this `if` block, filling in the order-type specific bits, and then fill in all the common parts afterwards.

Reply:
fixed